libsrc/leddevice/dev_spi/LedDeviceSK9822.cpp

```diff
@@ -0,0 +1,139 @@
+		_globalBrightnessControlMaxLevel = deviceConfig["globalBrightnessControlMaxLevel"].toInt(SK9222_GBC_MAX_LEVEL);
+		Info(_log, "[SK9822] Using global brightness control with threshold of %d and max level of %d", _globalBrightnessControlThreshold, _globalBrightnessControlMaxLevel);
+
+		const unsigned int startFrameSize = 4;
+		const unsigned int endFrameSize = qMax<unsigned int>(((_ledCount + 15) / 16), 4);
+		const unsigned int bufferSize = (_ledCount * 4) + startFrameSize + endFrameSize;
+
+		_ledBuffer.resize(bufferSize, 0xFF);
+		_ledBuffer[0] = 0x00;
+		_ledBuffer[1] = 0x00;
+		_ledBuffer[2] = 0x00;
+		_ledBuffer[3] = 0x00;
+
+		isInitOK = true;
+	}
+	return isInitOK;
+}
+
+
+void LedDeviceSK9822::bufferWithMaxCurrent(std::vector<uint8_t> &txBuf, const std::vector<ColorRgb> & ledValues, const int maxLevel) {
+	const int ledCount = static_cast<int>(_ledCount);
+
+	for (int iLed = 0; iLed < ledCount; ++iLed)
+	{
+		const ColorRgb &rgb = ledValues[iLed];
+		const uint8_t red = rgb.red;
+		const uint8_t green = rgb.green;
+		const uint8_t blue = rgb.blue;
+
+		/// The LED index in the buffer
+		const int b = 4 + iLed * 4;
+
+		// Use 0/31 LED-Current for Black, and full LED-Current for all other colors,
+		// with PWM control on RGB-Channels
+		const int ored = (red|green|blue);
+
+		txBuf[b + 0] = ((ored > 0) * (maxLevel & SK9222_GBC_MAX_LEVEL)) | SK9222_GBC_UPPER_BITS; // (ored > 0) is 1 for any r,g,b > 0, 0 otherwise; branch free
+		txBuf[b + 1] = red;
+		txBuf[b + 2] = green;
+		txBuf[b + 3] = blue;
+	}
+}
+
+void LedDeviceSK9822::bufferWithAdjustedCurrent(std::vector<uint8_t> &txBuf, const std::vector<ColorRgb> & ledValues, const int threshold, const int maxLevel) {
+	const int ledCount = static_cast<int>(_ledCount);
+
+	for (int iLed = 0; iLed < ledCount; ++iLed)
+	{
+		const ColorRgb &rgb = ledValues[iLed];
+		uint8_t red = rgb.red;
+		uint8_t green = rgb.green;
+		uint8_t blue = rgb.blue;
+		uint8_t level;
+
+		/// The LED index in the buffer
+		const int b = 4 + iLed * 4;
+
+		/// The maximal r,g,b-channel grayscale value of the LED
+		const uint16_t /* expand to 16 bit! */ maxValue = std::max(std::max(red, green), blue);
+
+		if (maxValue == 0) {
+			// Use 0/31 LED-Current for Black
+			level = 0;
+			red = 0x00;
+			green = 0x00;
+			blue = 0x00;
+		} else if (maxValue >= threshold) {
+			// Use full LED-Current when maximal r,g,b-channel grayscale value >= threshold and just use PWM control
+			level = (maxLevel & SK9222_GBC_MAX_LEVEL);
+		} else {
+			// Use adjusted LED-Current for other r,g,b-channel grayscale values
+			// See also: https://github.com/FastLED/FastLED/issues/656
+
+			// Scale the r,g,b-channel grayscale values to adjusted current = brightness level
+			const uint16_t /* 16 bit! */ brightness = (((maxValue + 1) * maxLevel - 1) >> 8) + 1;
+			#define scale(value) (((maxLevel * value + (brightness >> 1)) / brightness))
```

Comment:
No "define"s please. Use a C++ equivalent.
Thank you!

Reply:
Thanks for your feedback! This is now addressed in aa1c9e13191aa94c9972e4564525270c0606b79e .